### PR TITLE
feat(optimizer): Common sub-plan detection.

### LINKED ISSUE #7360: improve common sub-plan detection
see a example case at https://github.com/risingwavelabs/risingwave/issues/7343#issuecomment-1380040300

cc: @chenzl25 , feel free to reassign, thanks

### PR DESCRIPTION
**Pull request by wsx-ucb**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Attempt to resolve #7360.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>
#6955

### CODE REVIEW

**Comment by wsx-ucb** [comment]
The current design requires `PlanRef` to implement `Eq + Hash`.
The code currently does not compile, and there are some issues I would like to discuss:
1. Simply requiring `Eq + Hash` on `PlanNode` violates the object safety rules of Rust.
As a workaround, `DynEq` and `DynHash` are introduced as auxillary traits to achive object safety.
2. Implementing `Eq + Hash` for all plan nodes requires deriving those traits for almost all data structure involved.
But some data structures use `HashMap`, which has no implementation of `Hash`, and I restore to `BTreeMap` when possible to workaround the issue.
3. The `Share` nodes seems to insist on using `RefCell`, which entails interior mutability and has no implementation of `Hash` as well.
Are there any reason why such usage is encourage in the otherwise immutable plan node design?

**Comment by chenzl25** [comment]
> 3. The `Share` nodes seems to insist on using `RefCell`, which entails interior mutability and has no implementation of `Hash` as well.
>    Are there any reason why such usage is encourage in the otherwise immutable plan node design?

We have a method `clone_with_inputs` for each plan node, while for `Share` node, if we clone a new one, it will break `DAG` into `Tree` again. To work around , we use `replace_input` instead.

File: src/frontend/planner_test/tests/testdata/tpch.yaml [hunk 1]
```diff
@@ -2000,19 +2000,21 @@
     LogicalJoin { type: Inner, on: (sum($expr1) > $expr3), output: [partsupp.ps_partkey, sum($expr1)] }
     ├─LogicalAgg { group_key: [partsupp.ps_partkey], aggs: [sum($expr1), sum($expr1)] }
     | └─LogicalProject { exprs: [partsupp.ps_partkey, (partsupp.ps_supplycost * partsupp.ps_availqty) as $expr1] }
-    |   └─LogicalJoin { type: Inner, on: (supplier.s_nationkey = nation.n_nationkey), output: [partsupp.ps_partkey, partsupp.ps_availqty, partsupp.ps_supplycost] }
-    |     ├─LogicalJoin { type: Inner, on: (partsupp.ps_suppkey = supplier.s_suppkey), output: [partsupp.ps_partkey, partsupp.ps_availqty, partsupp.ps_supplycost, supplier.s_nationkey] }
-    |     | ├─LogicalScan { table: partsupp, columns: [partsupp.ps_partkey, partsupp.ps_suppkey, partsupp.ps_availqty, partsupp.ps_supplycost] }
-    |     | └─LogicalScan { table: supplier, columns: [supplier.s_suppkey, supplier.s_nationkey] }
-    |     └─LogicalScan { table: nation, output_columns: [nation.n_nationkey], required_columns: [nation.n_nationkey, nation.n_name], predicate: (nation.n_name = 'ARGENTINA':Varchar) }
+    |   └─LogicalShare { id = 373 }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/tpch.yaml:2003]: TPCH Q11 also benefits from this PR.

File: src/frontend/planner_test/tests/testdata/union.yaml [hunk 2]
```diff
@@ -206,9 +206,15 @@
 - sql: |
     select 1 union all select 1
   optimized_logical_plan: |
-    LogicalValues { rows: [[1:Int32], [1:Int32]], schema: Schema { fields: [1:Int32:Int32] } }
+    LogicalUnion { all: true }
+    ├─LogicalShare { id = 90 }
+    | └─LogicalValues { rows: [[1:Int32]], schema: Schema { fields: [1:Int32:Int32] } }
+    └─LogicalShare { id = 90 }
+      └─LogicalValues { rows: [[1:Int32]], schema: Schema { fields: [1:Int32:Int32] } }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/union.yaml:213]: Values is not worth to share.

File: src/frontend/planner_test/tests/testdata/expr.yaml [hunk 3]
```diff
@@ -472,9 +472,13 @@
...
       | └─StreamExchange { dist: Broadcast }
-      |   └─StreamNow { output: [now] }
+      |   └─StreamProject { exprs: [now], watermark_columns: [now] }
+      |     └─StreamShare { id = 94 }
+      |       └─StreamNow { output: [now] }
       └─StreamExchange { dist: Broadcast }
-        └─StreamNow { output: [now] }
+        └─StreamProject { exprs: [now], watermark_columns: [now] }
+          └─StreamShare { id = 94 }
+            └─StreamNow { output: [now] }
```
Comment (wsx-ucb) [inline src/frontend/planner_test/tests/testdata/expr.yaml:481]: We shouldn't share Now as well right?

Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/expr.yaml:481]: We can share the common sub-plan with the following properties:
1. include at least one table scan.
2. Not just simply a table scan.

**Comment by chenzl25** [comment]
Another interesting observation on nexmark q7. We can share more than just `HOP` window.

```
explain (trace) create materialized view v as SELECT AuctionBids.auction, AuctionBids.num FROM (
      SELECT
        bid.auction,
        count(*) AS num,
        window_start AS starttime
      FROM
        HOP(bid, date_time, INTERVAL '2' SECOND, INTERVAL '10' SECOND)
      GROUP BY
        window_start,
        bid.auction
    ) AS AuctionBids
    JOIN (
      SELECT
        max(CountBids.num) AS maxn,
        CountBids.starttime_c
      FROM (
        SELECT
          count(*) AS num,
          window_start AS starttime_c
        FROM HOP(bid, date_time, INTERVAL '2' SECOND, INTERVAL '10' SECOND)
        GROUP BY
          bid.auction,
          window_start
      ) AS CountBids
      GROUP BY
        CountBids.starttime_c
    ) AS MaxBids
    ON AuctionBids.starttime = MaxBids.starttime_c AND AuctionBids.num >= MaxBids.maxn;
```

This snippet can be shared as well. we just need to reorder its `group by` to ensure they are in the same order.
```sql
SELECT
        bid.auction,
        count(*) AS num,
        window_start AS starttime
      FROM
        HOP(bid, date_time, INTERVAL '2' SECOND, INTERVAL '10' SECOND)
      GROUP BY
        window_start,
        bid.auction
```

**Review by chenzl25 — APPROVED** [review]
LGTM! Great work!

**Review by st1page — APPROVED** [review]

**Comment by chenzl25** [comment]
> The failure should have nothing to do with this PR. It's a known issue of Sqlsmith fuzzy testing. Could someone force merge it after conflicts are resolved?

Too many conflicts. 🥵

**Pull request merged**